crates/bevy_reflect/src/path/access.rs

```diff
@@ -0,0 +1,240 @@
+use std::fmt;
+
+use super::ReflectPathError;
+use crate::{Reflect, ReflectMut, ReflectRef, VariantType};
+use thiserror::Error;
+
+type InnerResult<T> = Result<Option<T>, AccessError<'static>>;
+
+#[derive(Debug, PartialEq, Eq, Error)]
+pub enum AccessError<'a> {
+    #[error(
+        "the current {ty} doesn't have the {} {}",
+        access.kind(),
+        access.display_value(),
+    )]
+    Access { ty: Type, access: AccessRef<'a> },
+
+    #[error("invalid type shape: expected {expected} but found a reflect {actual}")]
+    Type { expected: Type, actual: Type },
+
+    #[error("invalid enum access: expected {expected} variant but found {actual} variant")]
+    Enum { expected: Type, actual: Type },
+}
+
+impl<'a> AccessError<'a> {
+    fn with_offset(self, offset: usize) -> ReflectPathError<'a> {
+        let error = self;
+        ReflectPathError::InvalidAccess { offset, error }
+    }
+}
+impl AccessError<'static> {
+    fn bad_enum(expected: Type, actual: impl Into<Type>) -> Self {
+        let actual = actual.into();
+        AccessError::Enum { expected, actual }
+    }
+    fn bad_type(expected: Type, actual: impl Into<Type>) -> Self {
+        let actual = actual.into();
+        AccessError::Type { expected, actual }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Type {
+    Struct,
+    TupleStruct,
+    Tuple,
+    List,
+    Array,
+    Map,
+    Enum,
+    Value,
+    Unit,
+}
+
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let name = match self {
+            Type::Struct => "struct",
+            Type::TupleStruct => "tuple struct",
+            Type::Tuple => "tuple",
+            Type::List => "list",
+            Type::Array => "array",
+            Type::Map => "map",
+            Type::Enum => "enum",
+            Type::Value => "value",
+            Type::Unit => "unit",
+        };
+        write!(f, "{name}")
+    }
+}
+impl<'a> From<ReflectRef<'a>> for Type {
+    fn from(value: ReflectRef<'a>) -> Self {
+        match value {
+            ReflectRef::Struct(_) => Type::Struct,
+            ReflectRef::TupleStruct(_) => Type::TupleStruct,
+            ReflectRef::Tuple(_) => Type::Tuple,
+            ReflectRef::List(_) => Type::List,
+            ReflectRef::Array(_) => Type::Array,
+            ReflectRef::Map(_) => Type::Map,
+            ReflectRef::Enum(_) => Type::Enum,
+            ReflectRef::Value(_) => Type::Value,
+        }
+    }
+}
+impl From<VariantType> for Type {
+    fn from(value: VariantType) -> Self {
+        match value {
+            VariantType::Struct => Type::Struct,
+            VariantType::Tuple => Type::Tuple,
+            VariantType::Unit => Type::Unit,
+        }
+    }
+}
+
+/// A singular owned element access within a path.
```

Comment:
This isn't very approachable to someone new to this section of the code base: can we try to explain in more detail / using synonyms what an "access" is?